### ISSUE TITLE
How does the pytorch seed affect the model in eval mode? 

### ISSUE THREAD

**Issue by ManuelEberhardinger** [facebookresearch/dinov2#323]
Hello, 

Thanks for providing the code and the models!

I have a question regarding the model in eval mode. We trained a model on our own custom dataset and then wanted to check how good it is for image retrieval. To do this, we loaded the full final checkpoint, saved from the Training Checkpointer, and created embeddings for our images. We then observed that these embeddings changed after reloading the checkpoint.

We checked the scripts in the eval folder and saw that the seeds are fixed for the evaluation of the model. We then loaded the model and only saved the state dict, as explained [here]( https://pytorch.org/tutorials/beginner/saving_loading_models.html#saving-loading-model-for-inference). After reloading the model from the saved file, it appears that the seed is fixed and results in the same embedding. We assumed that the seed was initialized randomly and that this is also stored in the state dict. However, for the checkpoint from the training, the seed is somehow randomly initialized each time the model is loaded. 

After a  reload the cosine similarity between embeddings for the same image drops from 1.0 to somewhere between +0.3 and -0.3. So the embeddings are totally different with a different seed. 

So I have two questions:
1. Why is the model still dependent on the seed in eval mode? 
2. Do different seeds influence the performance of the model after training is completed or do all seeds lead to similar results?

Thanks in advance!

Best,
Manuel

**Comment by qasfb** [comment]
This is unexpected, can you show your code ?

**Comment by ManuelEberhardinger** [comment]
I added a custom dataset, which is similar to the Imagenet class:
````+from .extended import ExtendedVisionDataset
+from typing import Any, Callable, Optional, Tuple
+from PIL import Image
+from pathlib import Path
+
+
+class Dokiq(ExtendedVisionDataset):
+
+    def __init__(
+        self,
+        *,
+        root: str = '/var/dokiq/dev/data-folder/file-data/thumbnails/medium',
+        transforms: Optional[Callable] = None,
+        transform: Optional[Callable] = None,
+        target_transform: Optional[Callable] = None,
+    ) -> None:
+        super().__init__(root, transforms, transform, target_transform)
+        path = Path(root)
+        self.paths = list(path.glob('**/*.png'))
+
+    def get_image_data(self, index: int) -> bytes:
+        image_path = self.paths[index]
+        image = Image.open(image_path).convert(mode="RGB")
+        return image
+
+    def get_target(self, index: int) -> Any:
+        return 0  # we return zero as we only have one class
+
+    def __getitem__(self, index: int) -> Tuple[Any, Any]:
+        try:
+            image = self.get_image_data(index)
+        except Exception as e:
+            raise RuntimeError(f"can not read image for sample {index}") from e
+        target = self.get_target(index)
+
+        if self.transforms is not None:
+            image, target = self.transforms(image, target)
+
+        return image, target
+
+    def __len__(self) -> int:
+        return len(self.paths)
````
I also updated the config vitl16_short.yaml to use the custom dataset: 
```` 
train:
  batch_size_per_gpu: 64
  dataset_path: Dokiq
student:
  block_chunks: 4
  ````
I ran this command to train the model: `PYTHONPATH=. python dinov2/run/train/train.py  --nodes 1  --gpus 1  --config-file dinov2/configs/train/vitl16_short.yaml --output-dir output_dokiq` 

to test the model we used this class:
````
class ImageEncoder(torch.nn.Module):
    checkpoint_path = Path('../output_dokiq/model_final.pth') # this model is from the checkpointer after dino training is finished
    
    #crop_size = 518
    #patch_size = 14
    default_config = {
        # from vit_large in models.py from the dino repo
        "embed_dim": 1024,
        "depth": 24,
        "num_heads": 16,
        "mlp_ratio": 4,
        "block_fn": partial(NestedTensorBlock, attn_class=MemEffAttention),

        # from hubconf.py in the dino repo
        "patch_size": 16,
        "img_size": 224,
        "ffn_layer": "mlp",
        "block_chunks": 4,
        "init_values": 1.0,
    }
    model_class = DinoVisionTransformer

    _instance: Optional["ImageEncoder"] = None

    preprocess_image = T.Compose([
        T.Lambda(lambda img: img.convert('RGB')),
        T.Resize(224),
        T.CenterCrop(224),
        T.ToTensor(),
        T.Normalize((0.48145466, 0.4578275, 0.40821073), (0.26862954, 0.26130258, 0.27577711))
    ])
    model_class: Type[torch.nn.Module]

    @classmethod
    def get(cls) -> "ImageEncoder":
        if cls._instance is None:
            cls._instance = cls.load_from_checkpoint(
                cls.checkpoint_path,
                config=cls.default_config
            )
            cls._instance.train(mode=False)
        return cls._instance

    def __init__(self, config):
        super().__init__()
        self.model = self.model_class(**config)
        self.model.eval()

    @classmethod
    def load_from_checkpoint(cls, checkpoint_path: Path, config: dict[str, Any]):
        dinov2_utils.fix_random_seeds(42) # without this line we always get different embeddings after each reload
        model = cls(config)
        dinov2_utils.load_pretrained_weights(model.model, str(checkpoint_path), "teacher")
        model.model.eval()
        return model

    def forward(self, x) -> torch.Tensor:
        return self.model.forward(x)

    def encode_images(self, image_batch: list[Image]) -> npt.NDArray[np.float32]:
        with torch.no_grad():
            preprocessed_images = torch.stack([self.preprocess_image(image) for image in image_batch]).cuda()
            image_features = self.forward(preprocessed_images)
            image_features /= image_features.norm(dim=-1, keepdim=True)
        return image_features

    def encode_image(self, image: Image) -> npt.NDArray[np.float32]:
        return self.encode_images([image])[0]
      
````
To check the similarity of embeddings, I just run: 
````
model = ImageEncoder.get()
model = model.cuda()
model = model.eval()

embedding = np.load('embedding.npy') # from the same model, but before the reload
img = Image.open("img")
new_embbeding = model.encode_image(img)
np.save('embedding.npy', new_embbeding)
new_embbeding = new_embbeding.T.cpu().numpy()
print(embedding @ new_embbeding) 
````

When I run `torch.save(model.model.state_dict(), 'model.pt')` the embeddings of the model will stay the same even after reloading the model. (Which I had also expected from using the model saved from the checkpointer)

**Comment by qasfb** [comment]
can you try to use the teacher_checkpoint.pth model from inside the eval/training_*** folder ?

**Comment by ManuelEberhardinger** [comment]
Thank you very much! It looks like that was the problem. Now the embeddings don't change after a reload and I don't have to fix the seeds either.

**Comment by qasfb** [comment]
I'm not sure what's going on; in .eval() mode things should do the same from one run to the other...
